### PR TITLE
openapi schema in CRD, validation tool, and auto-generation of doc for examples

### PR DESCRIPTION
**Pull request by jmazzitelli**
same as https://github.com/kiali/kiali-operator/pull/469 so please read that PR description for the motivation.

This PR includes the same validation script but it also includes descriptions in the CRD schema and a make target that can be used to [generate a .md file](https://github.com/giantswarm/crd-docs-generator) containing documentation for use in a hugo-generated website (such as the kiali.io website). The documentation can be used in place of the existing example kiali_cr.yaml, as an example kiali CR will be included in the documentation.

### CODE REVIEW

**Comment by jmazzitelli** [comment]
To see the validator still works, run it against the example CR:

```
crd-docs/config/crd/validate-kiali-cr.sh -kcf crd-docs/docs/cr/kiali.io_v1alpha1_kiali.yaml
```

results in:

```
Kiali CR [crd-docs/docs/cr/kiali.io_v1alpha1_kiali.yaml] is valid.
```

**Comment by jmazzitelli** [comment]
the CRD schema now has all the same help text that the example kiali cr had. Auto-generating the CRD schema doc page will have the same information.

**Comment by jmazzitelli** [comment]
To see what the new doc page will look like (Assuming your kiali.io repo is in `~/source/kiali/kiali/io` - put your path in there if different):
```
make clean gen-crd-doc && cp _output/crd-docs/kialis.kiali.io.md ~/source/kiali/kiali.io/content/en/docs/Configuration
```
Now in the kiali.io repo, run "make serve" and go to http://localhost:1313/docs/configuration/kialis.kiali.io/ and you will see the new auto-generated page. It shows you an example CR (which we still need to manually update - but it is just the values, no comments) and it shows you the CRD schema descriptions of all the fields (we need to maintain that CRD schema). That CRD schema is also used by the validation script so one CRD file and we (a) have a dev tool for validation and (b) have a source of truth for the auto-generated docs.

**Comment by jmazzitelli** [comment]
One thing that has to change - either right before we merge we need to change (and then reapprove) this - or we submit a followup PR with this change:

Go from this (which is needed to test this PR because my repo branch is the only place where the crd currently is):
```yaml
source_repositories:
- url: https://github.com/jmazzitelli/kiali-operator
  organization: jmazzitelli
  short_name: kiali-operator
  commit_reference: openapi-crd-generate-api-docs
```

To this (which points to the kiali master branch which is needed when we start producing the docs):

```yaml
source_repositories:
- url: https://github.com/kiali/kiali-operator
  organization: kiali
  short_name: kiali-operator
  commit_reference: master
```

**Comment by jmazzitelli** [comment]
I would like to delete the `config` and `docs` top level directories here because it just pollutes the repo a bit. But we need this enhancement request to the crd-docs-generation tool merged and released before we can do this - https://github.com/giantswarm/crd-docs-generator/pull/94

File: Makefile
Comment (nrfox) [inline Makefile:99]: Doesn't `kiali/kiali` already generate openapi docs using the `swagger` tool? Can the same tool be re-used here?

Comment (jmazzitelli) [reply Makefile:99]: AFAIK, swagger cannot parse a k8s CRD with OpenAPI3 schema which is what this tool does.

**Comment by jmazzitelli** [comment]
When https://github.com/giantswarm/crd-docs-generator/pull/95 is available, we can re-organize the files a little bit better. We won't need to have a "config/crd" and "docs/cr" subdirectories. We can put them more organized locations (something like "crd-docs/config", "crd-docs/crd", "crd-docs/example-cr").

**Comment by jmazzitelli** [comment]
> When [giantswarm/crd-docs-generator#95](https://github.com/giantswarm/crd-docs-generator/pull/95) is available, we can re-organize the files a little bit better. We won't need to have a "config/crd" and "docs/cr" subdirectories. We can put them more organized locations (something like "crd-docs/config", "crd-docs/crd", "crd-docs/example-cr").

Rather than wait for the release, I just published the new container on my own quay.io repo. I'll use that for now. When the new release is done, I'll update the Makefile so it runs the official release image, not the one I built and published on my own quay repo.

This code has been committed to this PR and is testable.

**Comment by jmazzitelli** [comment]
This PR is ready for review and merge. @nrfox I officially nominate you as reviewer ;) 

## How to test

1. `make validate-cr` confirms the example CR validates successfully against the CRD schema. Run it and it should complete successfully.
2. Edit `crd-docs/cr/kiali.io_v1alpha1_kiali.yaml` and introduce an error (e.g. add a bogus/invalid entry like `spec.foo = bar`). Re-run `make validate-cr` and confirm the failure message describes the error you introduced.
3. `make gen-crd-doc` confirms the documentation is successfully generated from the CRD schema; this replaces our example kiali_cr.yaml file with true markdown documentation we can add to our kiali.io doc website (in the `content/en/docs/Configuration` folder). After you run that make target, look in `_output/crd-docs` and confirm the existence and correctness of `_output/crd-docs/kialis.kiali.io.md`. You can more easily review that file if you copy it in your kiali.io repo's `content/en/docs/Configuration` directory and run `make serve` in the kiali.io and point your browser to the configuration documentation.

## Finishing touches needed

Because of the nature of how the doc gen tool works, the following changes need to be made before we merge (but you can't test this PR with these changes - which is why I haven't done them yet. After review is OK, I'll make one last commit, at which point we'll get one last official +1 review and then we can merge).

1. We need to point the gen tool to the master branch of the kiali repo:
```diff
diff --git a/crd-docs/config/apigen-config.yaml b/crd-docs/config/apigen-config.yaml
index 32fdfee..3157c73 100644
--- a/crd-docs/config/apigen-config.yaml
+++ b/crd-docs/config/apigen-config.yaml
@@ -2,9 +2,9 @@ template_path: ./apigen-crd.template
 
 source_repositories:
 - url: https://github.com/jmazzitelli/kiali-operator
-  organization: jmazzitelli
+  organization: kiali
   short_name: kiali-operator
-  commit_reference: openapi-crd-generate-api-docs
+  commit_reference: master
   crd_paths:
   - crd-docs/crd
   cr_paths:
```

2. If there is a [new release](https://quay.io/repository/giantswarm/crd-docs-generator?tab=tags) of the gen tool that contains [the enhancement](https://github.com/giantswarm/crd-docs-generator/pull/95) we need (which should be in anything after 0.8.0), we can make this change also. If the enhancement is not released yet, we'll have to do this in a followup PR when the release is available.
```diff
diff --git a/Makefile b/Makefile
index 8f13cc6..462f6b9 100644
--- a/Makefile
+++ b/Makefile
@@ -100,7 +100,7 @@ validate-cr:
 ## gen-crd-doc: Generates documentation for the Kiali CR configuration
 gen-crd-doc:
        mkdir -p ${OUTDIR}/crd-docs
-       ${DORP} run -v ${OUTDIR}/crd-docs:/opt/crd-docs-generator/output -v ${ROOTDIR}/crd-docs/config:/opt/crd-docs-generator/config quay.io/jmazzitelli/crd-docs-generator:latest --config /opt/crd-docs-generator/config/apigen-config.yaml
+       ${DORP} run -v ${OUTDIR}/crd-docs:/opt/crd-docs-generator/output -v ${ROOTDIR}/crd-docs/config:/opt/crd-docs-generator/config quay.io/giantswarm/crd-docs-generator:latest --config /opt/crd-docs-generator/config/apigen-config.yaml
 
 # Ensure "docker buildx" is available and enabled. For more details, see: https://github.com/docker/buildx/blob/master/README.md
 # This does a few things:
```

File: crd-docs/bin/validate-kiali-cr.sh
Comment (nrfox) [inline crd-docs/bin/validate-kiali-cr.sh:140]: I wish there was a way to validate a CR against a CRD's openapi schema offline. I'm a little leery of creating/deleting a new CRD and CR to validate. If an error happens before cleanup you could end up with two kiali CRDs installed in your cluster and unknowingly use the testing one. There doesn't seem to be a way to do offline validation with the standard tooling though. And this is mostly a diagnostic tool anyways.

**Review by nrfox — APPROVED** [review]

File: crd-docs/bin/validate-kiali-cr.sh
Comment (jmazzitelli) [reply crd-docs/bin/validate-kiali-cr.sh:140]: I agree with all of the above. And I did think about the case where, for some reason, the test CRD fails to delete. It will be annoying because you will need to manually delete it. However, do note - the test CRD that is created by this tool is literally called "TestKiali" - it is NOT the same name as the CRD that the operator/helm will create and is NOT the CRD that will be used by the Kiali operator itself. So, yes, if the test CRD fails to delete there will be two CRDs, but the TestKiali CRD is essentially unused. The operator ignores it completely. You can create CRs of the kind "TestKiali" (which is what this tool does) and those objects do nothing - they will just float around without any impact to Kiali itself.

**Comment by jmazzitelli** [comment]
@nrfox The "finishing touches" commit is in - just need a final +1 approval and I'll merge.

**Review by nrfox — APPROVED** [review]

**Pull request merged**
